pkg/integrations/v4/dm/emitter.go

```diff
@@ -0,0 +1,274 @@
+}
+
+func (e *emitter) RegisterEntities(entities []protocol.Entity) (registeredEntitiesNameToID, unregisteredEntityListWithWait) {
+	// Bulk update them (after checking our datastore if they exist)
+	// add entity ID to metric annotations
+	return e.idProvider.ResolveEntities(entities)
+}
+
+func emitInventory(
+	emitter agent.PluginEmitter,
+	metadata integration.Definition,
+	integrationMetadata protocol.IntegrationMetadata,
+	entityID entity.ID,
+	dataSet protocol.Dataset,
+	labels map[string]string) {
+	logEntry := elog.WithField("action", "EmitV4DataSet")
+
+	integrationUser := metadata.ExecutorConfig.User
+
+	if len(dataSet.Inventory) > 0 {
+		inventoryDataSet := legacy.BuildInventoryDataSet(
+			logEntry, dataSet.Inventory, labels, integrationUser, integrationMetadata.Name,
+			dataSet.Entity.Name)
+		entityKey := entity.Key(dataSet.Entity.Name)
+		emitter.EmitInventory(inventoryDataSet, entity.New(entityKey, entityID))
+	}
+}
+
+func emitEvent(
+	emitter agent.PluginEmitter,
+	metadata integration.Definition,
+	dataSet protocol.Dataset,
+	labels map[string]string) {
+
+	integrationUser := metadata.ExecutorConfig.User
+	for _, event := range dataSet.Events {
+		normalizedEvent := legacy.
+			NormalizeEvent(elog, event, labels, integrationUser, dataSet.Entity.Name)
+		if normalizedEvent != nil {
+			emitter.EmitEvent(normalizedEvent, entity.Key(dataSet.Entity.Name))
+		}
+	}
+}
+
+// Replace entity name by applying entity rewrites and replacing loopback
+func replaceEntityName(entity protocol.Entity, entityRewrite []data.EntityRewrite, agentShortName string) {
+	newName := legacy.ApplyEntityRewrite(entity.Name, entityRewrite)
+	newName = http.ReplaceLocalhost(newName, agentShortName)
+	entity.Name = newName
+}
+
+// ParsePayloadV4 parses a string containing a JSON payload with the format of our
+// SDK for v4 protocol which uses dimensional metrics.
+func ParsePayloadV4(raw []byte, ffManager feature_flags.Retriever) (dataV4 protocol.DataV4, err error) {
+	if len(raw) == 0 {
+		err = NoContentToParseErr
+		return
+	}
+
+	if enabled, ok := ffManager.GetFeatureFlag(handler.FlagProtocolV4); !ok || !enabled {
+		err = ProtocolV4NotEnabledErr
+		return
+	}
+
+	err = json.Unmarshal(raw, &dataV4)
+	return
+}
+
+// Returns a composed error which describes all the errors found during the emit process of each data set
+func composeEmitError(emitErrs []error, dataSetLenght int) error {
+	if len(emitErrs) == 0 {
+		return nil
+	}
+
+	composedError := fmt.Sprintf("%d out of %d datasets could not be emitted. Reasons: ", len(emitErrs), dataSetLenght)
+	messages := map[string]struct{}{}
+
+	for _, err := range emitErrs {
+		msg := err.Error()
+		if _, ok := messages[msg]; !ok { // avoid logging repeated error messages
+			messages[msg] = struct{}{}
+			composedError += msg + ","
+		}
+	}
+	return errors.New(composedError[:len(composedError)-1])
+}
+
+func wrapError(err error, datasetLen int) error {
```

Comment:
hey, `%d out of %d` prints out the same value

Reply:
sorry, I do not understand what is it about. Can you give me more context about you thought?

Reply:
the line below: 	

composedError := fmt.Sprintf("%d out of %d datasets could not be emitted. Reasons: %v", **datasetLen, datasetLen**, err)